spec/datadog/profiling/stack_recorder_spec.rb

```diff
@@ -302,6 +336,88 @@ def sample_types_from(decoded_profile)
+
+      let(:a_string) { 'a beautiful string' }
+      let(:an_array) { [1..10] }
+      let(:a_hash) { { 'a' => 1, 'b' => '2', 'c' => true } }
+
+      let(:samples) { samples_from_pprof(encoded_pprof) }
+
+      before do
+        allocations = [a_string, an_array, 'a fearsome string', [-10..-1], a_hash, { 'z' => -1, 'y' => '-2', 'x' => false }]
+        @num_allocations = 0
+        allocations.each_with_index do |obj, i|
+          # Heap sampling currently requires this 2-step process to first pass data about the allocated object...
+          described_class::Testing._native_track_object(stack_recorder, obj, sample_rate)
+          # ...and then pass data about the allocation stacktrace (with 2 distinct stacktraces)
+          if i.even?
+            Datadog::Profiling::Collectors::Stack::Testing
+              ._native_sample(Thread.current, stack_recorder, metric_values1, labels, numeric_labels, 400, false)
+          else
+            Datadog::Profiling::Collectors::Stack::Testing
+              ._native_sample(Thread.current, stack_recorder, metric_values2, labels, numeric_labels, 400, false)
+          end
+          @num_allocations += 1
+        end
+
+        allocations.clear # The literals in the previous array are now dangling
+        GC.start # And this will clear them, leaving only the non-literals which are still pointed to by the lets
+      end
+
+      context 'when disabled' do
+        let(:heap_samples_enabled) { false }
+
+        it 'are ommitted from the profile' do
+          # We sample from 2 distinct locations
+          expect(samples.size).to eq(2)
+          expect(samples.select { |s| s.values.key?('heap-live-samples') }).to eq([])
+        end
+      end
+
+      context 'when enabled' do
+        let(:heap_samples_enabled) { true }
+
+        let(:heap_samples) do
+          samples.select { |s| s.values[:'heap-live-samples'] > 0 }
+        end
+
+        let(:non_heap_samples) do
+          samples.select { |s| s.values[:'heap-live-samples'] == 0 }
+        end
+
+        it 'are included in the profile' do
+          pending 'heap_recorder implementation is currently missing'
+
+          # We sample from 2 distinct locations
+          expect(heap_samples.size).to eq(2)
+
+          sum_heap_samples = 0
+          heap_samples.each { |s| sum_heap_samples += s.values[:'heap-live-samples'] }
+          expect(sum_heap_samples).to eq([a_string, an_array, a_hash].size * sample_rate)
+        end
+
+        it 'do not corrupt/overwrite non-heap-samples' do
+          expect(non_heap_samples.size).to eq(2)
+
+          sum_cpu_time = 0
+          sum_alloc_samples = 0
+
+          non_heap_samples.each do |s|
+            sum_cpu_time += s.values[:'cpu-time']
+            sum_alloc_samples += s.values[:'alloc-samples']
+          end
+
+          expect(sum_cpu_time).to be > 0
+          expect(sum_alloc_samples).to eq(@num_allocations * sample_rate)
+        end
```

Comment:
I tend to agree with the intention in general. But in this case, because of the extra work needed to "mock" allocation tracking in these tests, this section is the only one where we actually have `heap_samples_enabled`.

So all the other tests in this suite that validate non-heap profiling do not cover `heap_samples_enabled=true` case which is also the only case where we do "out-of-band" sample writing that could, in theory, overwrite/mess with something that happens in the normal recording. The intention of this test is to serve as a canary for such a problem.

An alternative might be adding a positive test case with `heap_samples_enabled=true` to the existing test context?